pkgs/development/libraries/qt-5/5.15/default.nix

```diff
@@ -288,6 +289,12 @@ let
       qtwayland = callPackage ../modules/qtwayland.nix {};
       qtwebchannel = callPackage ../modules/qtwebchannel.nix {};
       qtwebengine = callPackage ../modules/qtwebengine.nix {
+        # The version of Chromium used by Qt WebEngine 5.15.x does not build with clang 16 due to
+        # -Wenum-constexpr-conversion errors in v8 and incompatible function pointer conversion
+        # errors in harfbuzz-ng. Pinning the version to clang 15 allows it to build.
```

Comment:
Your github comment says

> The HarfBuzz error is fixable, but the v8 one is not.

I assume you mean "is not fixed by disabling warnings".  If so, then it the problem won't get better with future versions of clang, and qtwebengine-5.x is not getting any more updates to its chromium, so I guess we don't have to worry about the pin getting stale.

Reply:
The error v8 is triggering can be suppressed in clang 16 but will (probably) be made a hard error in a future version of clang (because it’s UB in C++17). See llvm/llvm-project#59036.

Someone who is savvy enough might be able to fix it, but my C++ is pretty rusty. After attempting a backport for v8 in nodejs_18 and failing, the safest route is to pin the clang version.

Reply:
The HardBuzz issue is patched upstream, and porting the patch is trivial. It’s just doesn’t seem worth doing when the patch isn’t needed for clang 15.

Reply:
I'm having trouble following this conversation because you're using the term "fixing" for both "disable warnings" and "heavily patch upstream's code".

Let's try to separate these.

If you can get this to build on clang 16 by disabling warnings, you should absolutely do that instead of pinning clang.

It would help if your comment distinguished between "disabling warnings" and "large patches" instead of lumping them together under the word "fix".

Reply:
I’ll clarify the comment, but clang will remain pinned at clang 15 (matching top-level v8 and nodejs_18).

The clang developers have indicated that `-Wenum-constexpr-conversion` will be turned into a hard error in the future. It’s not certain yet whether that will happen for clang 18, but it is certain that Darwin will be updating to clang 18 for 24.11. Pinning clang now avoids having to deal with the breakage in the future.

Reply:
I clarified the comment and the commit message. I also rebased to address the merge conflict. Let me know if there’s anything else I can do better to explain the situation and the chosen solution.

Reply:
I should also add that I did try to build it with clang 16. After backporting the patches for HarfBuzz, I found that v8 was failing to compile for the same reasons that nodejs_18 (and older) and v8 were failing to compile. That’s what prompted me to take the same approach as was taken to fix those build failures.

One of my assumptions is the clang developers mean it when they this will become a hard error in the future. Diagnosing and fixing these problems is the most time-consuming part of a default clang update, so if I can avoid having to do it again, I’d like to take that approach.

If it turns out the clang do decide to make it a hard error only in C++17 mode and not one in C++14 and earlier, these pins can be revisited as long as compiling in C++14 mode works and does not have ABI implications.